editor/src/core/vscode/vscode-bridge.ts

```diff
@@ -0,0 +1,230 @@
+  isEditorCursorPositionChanged,
+} from 'utopia-vscode-common'
+import { isTextFile, ProjectFile } from '../shared/project-file-types'
+import { isBrowserEnvironment } from '../shared/utils'
+import { getTemplatePathsInBounds } from '../../components/editor/store/editor-state'
+
+const Scheme = 'utopia'
+const RootDir = `/${Scheme}`
+
+function toFSPath(projectPath: string): string {
+  const fsPath = appendToPath(RootDir, projectPath)
+  return fsPath
+}
+
+function fromFSPath(fsPath: string): string {
+  const prefix = RootDir
+  const prefixIndex = fsPath.indexOf(prefix)
+  if (prefixIndex === 0) {
+    const projectPath = fsPath.slice(prefix.length)
+    return projectPath
+  } else {
+    throw new Error(`Invalid FS path: ${fsPath}`)
+  }
+}
+
+async function writeProjectFile(
+  projectID: string,
+  projectPath: string,
+  file: ProjectFile,
+): Promise<void> {
+  switch (file.type) {
+    case 'DIRECTORY': {
+      return ensureDirectoryExists(toFSPath(projectPath))
+    }
+    case 'TEXT_FILE': {
+      return writeFileAsUTF8(toFSPath(projectPath), file.fileContents.code)
+    }
+    case 'ASSET_FILE':
+      return Promise.resolve()
+    case 'IMAGE_FILE':
+      return Promise.resolve()
+  }
+}
+
+async function writeProjectContents(
+  projectID: string,
+  projectContents: ProjectContentTreeRoot,
+): Promise<void> {
+  await walkContentsTreeAsync(projectContents, (fullPath, file) => {
+    if (isTextFile(file) || isDirectory(file)) {
+      return writeProjectFile(projectID, fullPath, file)
+    } else {
+      return Promise.resolve()
+    }
+  })
+}
+
+function watchForChanges(projectID: string, dispatch: EditorDispatch): void {
+  function onCreated(fsPath: string): void {
+    stat(fsPath).then((fsStat) => {
+      if (fsStat.type === 'FILE') {
+        readFileAsUTF8(fsPath).then((text) => {
+          const action = updateFromCodeEditor(fromFSPath(fsPath), text)
+          dispatch([action], 'everyone')
+        })
+      }
+    })
+  }
+  function onModified(fsPath: string): void {
+    onCreated(fsPath)
+  }
+  function onDeleted(fsPath: string): void {
+    const projectPath = fromFSPath(fsPath)
+    const action = deleteFile(projectPath)
+    dispatch([action], 'everyone')
+  }
+  watch(toFSPath('/'), true, onCreated, onModified, onDeleted)
+}
+
+export async function initVSCodeBridge(
+  projectID: string,
+  projectContents: ProjectContentTreeRoot,
+  dispatch: EditorDispatch,
+): Promise<void> {
+  if (isBrowserEnvironment) {
+    await initializeFS(projectID)
+    await clearBothMailboxes()
+    await writeProjectContents(projectID, projectContents)
+    initMailbox(UtopiaInbox, parseFromVSCodeMessage, (message: FromVSCodeMessage) => {
+      if (isEditorCursorPositionChanged(message)) {
+        dispatch(
+          [selectFromFileAndPosition(message.filePath, message.line, message.column)],
+          'everyone',
+        )
+      } else {
+        console.error(`Unhandled message type ${JSON.stringify(message)}`)
```

Comment:
same question about `const _exhaustiveCheck: never = message`

Reply:
We actually can't here because there is only one type for this message, so I've removed the `if` `else` statement since now if a new message type is added it will cause a compile error there instead now.